crates/common/tedge_utils/src/fs.rs

```diff
@@ -40,40 +40,36 @@ impl<T, E: Into<std::io::Error>> ErrContext<T> for Result<T, E> {
     }
 }
 
-pub struct TempFile(PathBuf);
-
-impl Drop for TempFile {
-    fn drop(&mut self) {
-        let _ = std_fs::remove_file(&self.0);
-    }
-}
-
 /// Write file to filesystem atomically using std::fs synchronously.
 pub fn atomically_write_file_sync(
     dest: impl AsRef<Path>,
     mut reader: impl Read,
 ) -> Result<(), AtomFileError> {
     let dest_dir = parent_dir(dest.as_ref());
-    // FIXME: `.with_extension` replaces file extension, so if we used this
-    // function to write files `file.txt`, `file.bin`, `file.jpg`, etc.
-    // concurrently, then this will result in an error
-    let tempfile = TempFile(PathBuf::from(dest.as_ref()).with_extension("tmp"));
-
-    // Write the content on a temp file
-    let mut file = std_fs::OpenOptions::new()
-        .write(true)
-        .create_new(true)
-        .open(&tempfile.0)
-        .with_context(
-            || format!("could not create the temporary file {:?}", tempfile.0),
-            &dest,
-        )?;
+
+    // removed on drop
+    let mut file = tempfile::Builder::new()
+        .rand_bytes(6)
+        // FIXME: setting tempfile permissions to 644 is required not to break config
+        //
+        // normally tempfile creates a temporary file with mode 600, which is good for tmpfiles but
+        // this function is used for saving /etc/tedge/tedge.toml, which doesnt explicitly set a
+        // mode, so it assumes file created by this function be with default umask and that default
+        // umask is 022; but in the test suite `bootstrap.sh` is run as root, which makes
+        // /etc/tedge/tedge.toml owned by root, but later when `tedge-mapper-c8y` runs as `tedge` it
+        // can no longer read the config file, rendering thin-edge completely unusable. the fix is
+        // not trivial, because we can't set permissions on the config calling side if
```

Comment:
I didn't get that. Why couldn't the config file be created with the desired mode (644)?

Reply:
Same for me. I don't understand the comment.

Reply:
The comment was prompted by 2 things:

1. this `TODO` comment:
    https://github.com/thin-edge/thin-edge.io/blob/a9f2fcfcd211230cf926521db5f2802beab1fcce/crates/common/tedge_config/src/tedge_config_cli/tedge_config_location.rs#L145-L156
    When in the test suite, `bootstrap.sh` is run as root, then this function is run as root and `/etc/tedge/tedge.toml` ends up being owned by root, so it needs `a+r` permissions to be readable. As the `TODO` says, we should probably instead set ownership and mode in this function. In e090dc0 this is fixed by setting permissions and warning on errors. Is this acceptable?
2. Other code that calls these functions implicitly assumed that resulting destination files would have permissions 644, but had the umask been different the permissions would be different and this assumption would be violated.
    Now the function manually sets 644 permissions and additional doc comment was written to clarify what happens with file permissions (e090dc0).

Perhaps I initially overestimated how hard it would be to handle setting the permissions in the `store` function because tedge-agent is run as either root or tedge-agent, but we can just ignore the error. As such `FIXME` comment was removed.